wgsl/index.bs

```diff
@@ -454,6 +454,149 @@ whether failure to meet a particular requirement
 results in a shader-creation, pipeline-creation, or dynamic error.
 
 The WebGPU specification describes the consequences of each kind of error.
+A WGSL program with a [=shader-creation error=] or [=pipeline-creation error=] error will not be incorporated
+into a [=pipeline=] and hence will not be executed.
+
+## Diagnostics ## {#diagnostics}
+
+An implementation can generate [=diagnostics=] during [=shader module creation=].
+A <dfn noexport>diagnostic</dfn> is a message produced by the implementation for the benefit of the application author.
+
+We say a diagnostic is <dfn noexport>triggered</dfn> when particular condition is met, known as the [=diagnostic/triggering rule=].
+The place in the source text where the condition is met is known as the <dfn dfn-for="diagnostic">triggering location</dfn>.
+
+A [=diagnostic=] has the following properties:
+* A [=diagnostic/severity=].
+* A [=diagnostic/triggering rule=].
+* A [=diagnostic/triggering location=], a source location where the triggering rule applies, expressed as a point or range in the source text.
+* Optionally, details describing how the [=diagnostic/triggering rule=] applies in this case.
+
+The <dfn dfn-for="diagnostic">severity</dfn> of a diagnostic is one of:
+: <dfn dfn-for="severity" noexport>error</a>
+:: The diagnostic an error.
+     This is the greatest severity.
+: <dfn dfn-for="severity" noexport>warning</a>
+:: The diagnostic describes an anomaly that merits the attention of the application developer, but is not an error.
+: <dfn dfn-for="severity" noexport>info</a>
+:: The diagnostic describes a notable condition that merits attention of the application developer, but is not an error or warning.
+    This is the lowest severity.
+
+Triggered diagnostics [=behavioral requirement|will=] be processed as follows:
+1. If a [=shader-creation error=] occurs independently of diagnostic processing, then all diagnostics may be discarded.
+2. Sort [=diagnostic filters=] so that DF1 precedes DF2 if the DF1's [=affected range=] includes DF2's affected range.
```

Comment:
The way filtering is described seems counter-intuitive to me. Imagine a situation where we have something like the following:
```
@diagnostic(off, derivative_uniformity) {
  let a = textureSample(...);
  let b = @diagnostic(error, derivative_uniformity) textureSample(...);
}
```
The outer diagnostic precedes the inner diagnostic so the error is discard on `b`'s initializer despite asking for an error. Perhaps this should be described as synthesizing a filter with a priority based on most precise location for any given diagnostic.

The current rules work for warning -> error, but the discard from off happens too early.

Reply:
You're right.  That makes it actually lexical, which is very intuitive, and matches the push/pop paradigm in other compilers.

I had thought there was a use case for the mixing, but I didn't think it through well enough.

Reply:
Redone to make lexcially scoped w.r.t. triggering location. Thanks!